### PR TITLE
Add max pool op (with index)

### PR DESCRIPTION
**Pull request by chengduoZH**
fix https://github.com/PaddlePaddle/Paddle/issues/4327

### CODE REVIEW

**Comment by chengduoZH** [comment]
I've commented off the check_grad in max pooling because the current method of testing can not effectively test the gradient of max pooling. But it ensures that the gradients computed by GPU and CPU are equal.

File: paddle/operators/pool_with_index_op.cc [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/operators/pool_with_index_op.h"
+
+namespace paddle {
+namespace operators {
+
+int OutputSizeMaxPool(int input_size, int filter_size, int padding,
```
Comment (typhoonzero) [inline paddle/operators/pool_with_index_op.cc]: This function can be a `inline`. And, this function can be of more general use for `conv` and `pool` operators.

File: paddle/operators/pool_with_index_op.cc [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+the input and ksize, strides, paddings parameters.
+)DOC");
+  }
+};
+}  // namespace operators
+}  // namespace paddle
+
+namespace ops = paddle::operators;
+
+REGISTER_OP(maxPool2dWithIndex, ops::MaxPoolWithIndexOp,
```
Comment (typhoonzero) [inline paddle/operators/pool_with_index_op.cc]: Seems operator names are commonly `xxx_xxx_op` but not "camal case".

File: paddle/operators/CMakeLists.txt [hunk 3]
```diff
@@ -75,6 +75,12 @@ function(op_library TARGET)
         file(APPEND ${pybind_file} "USE_OP(reduce_sum);\n")
     endif()
 
+    if ("${TARGET}" STREQUAL "pool_with_index_op")
+        set(pybind_flag 1)
+        # It's enough to just adding one operator to pybind
```
[the remark this answers is not in the document]
Comment (chengduoZH) [reply paddle/operators/CMakeLists.txt]: Done

File: paddle/operators/math/pooling.cc [hunk 4]
```diff
@@ -458,6 +458,233 @@ template class Pool3dGradFunctor<
     platform::CPUPlace, paddle::operators::math::MaxPoolGrad<double>, double>;
 template class Pool3dGradFunctor<
     platform::CPUPlace, paddle::operators::math::AvgPoolGrad<double>, double>;
+
+template <typename T>
+class MaxPool2dWithIndexFunctor<platform::CPUPlace, T> {
```
[the remark this answers is not in the document]
Comment (chengduoZH) [reply paddle/operators/math/pooling.cc]: Done. Thx !!!

File: paddle/operators/pool_with_index_op.cc (hunk 1 above)
Comment (chengduoZH) [reply paddle/operators/pool_with_index_op.cc]: Done.
This function is frequently called in conv2/3d, conv2/3d_cudnn, pool2/3d and pool2/3_cudnn. 
I think we should convert this function to functor, and move to namespace math.

File: paddle/operators/pool_with_index_op.cc (hunk 2 above)
Comment (chengduoZH) [reply paddle/operators/pool_with_index_op.cc]: Done.

File: paddle/operators/CMakeLists.txt [hunk 5]
```diff
@@ -75,6 +75,13 @@ function(op_library TARGET)
         file(APPEND ${pybind_file} "USE_OP(reduce_sum);\n")
     endif()
 
+    # pool_with_index_op contains several operators
+    if ("${TARGET}" STREQUAL "pool_with_index_op")
+        set(pybind_flag 1)
+        # It's enough to just adding one operator to pybind
+        file(APPEND ${pybind_file} "USE_OP(max_pool2d_with_index);\n")
```
Comment (typhoonzero) [inline paddle/operators/CMakeLists.txt]: Also need to add `max_pool3d_with_index`

File: paddle/operators/pool_with_index_op.cc [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(framework::InferShapeContextBase *ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("X"),
+                   "X(Input) of Pooling should not be null.");
+    PADDLE_ENFORCE(
+        ctx->HasOutput(framework::GradVarName("X")),
+        "X@GRAD(Input@GRAD) of MaxPoolWithIndexOpGrad should not be null.");
```
Comment (typhoonzero) [inline paddle/operators/pool_with_index_op.cc]: `@GRAD` suffix is for internal use, may confuse users if this message came out.

File: paddle/operators/pool_with_index_op.cc [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+              "The format of output tensor is also NCHW."
+              "Where N is batch size, C is the number of channels, H and W "
+              "is the height and width of image."
+              "The value in it is the index in current feature map");
+
+    AddAttr<std::vector<int>>(
+        "ksize",
+        "Pooling size(height, width) of pooling operator."
+        "If globalPooling = true, ksize is ignored and need not be "
+        "specified.");  // TODO(Add checker)
```
Comment (typhoonzero) [inline paddle/operators/pool_with_index_op.cc]: Seems you need to configure pre-commit on your local environment before running `git commit`, so the style check can automatically prints some style errors.

In this case, "TODO" comments should be like: https://google.github.io/styleguide/cppguide.html#TODO_Comments

File: paddle/operators/pool_with_index_op.cc [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput(
+        "X",
+        "The input tensor of pooling operator. "
+        "The format of input tensor is NCDHW. Where N is batch size, C is "
+        "the number of channels, D, H and W is the depth, height and width of "
+        "image.");
+    AddOutput("Out",
+              "The output tensor of pooling operator."
+              "The format of output tensor is also NCDHW."
```
Comment (typhoonzero) [inline paddle/operators/pool_with_index_op.cc]: > "The format of output tensor is also NCDHW."
"Where N is batch size, C is "
"the number of channels, D, H and W is the depth, height and "
"width of image."

Can put these explains to `AddComment`, so here can be more simpler.

File: paddle/operators/pool_with_index_op.cc (hunk 7 above)
Comment (chengduoZH) [reply paddle/operators/pool_with_index_op.cc]: Thanks, this style of writing is really nonstandard.
I have configured pre-commit on my local environment, but it also pass.

File: paddle/operators/pool_with_index_op.cc (hunk 6 above)
Comment (chengduoZH) [reply paddle/operators/pool_with_index_op.cc]: I approve with you. If X@GRAD is null, it must be an internet error. But I think it is helpful to developer.

File: paddle/operators/CMakeLists.txt (hunk 5 above)
Comment (chengduoZH) [reply paddle/operators/CMakeLists.txt]: Just adding one operator of pool_with_index_op is ok. And unit test(max_pool2d_with_index, max_pool3d_with_index) passed.

**Pull request merged**